transform_assumption.ml

```diff
@@ -1,37 +1,192 @@
+  let version = List.hd func.body in
+  let succ = Analysis.successors version.instrs in
+  let pred = Analysis.predecessors version.instrs in
+  if (not version.has_checkpoints) then None else
+  let transform pc =
+    match[@warning "-4"] version.instrs.(pc) with
+    | Osr {cond} when cond = [] ->
+      Remove 1
+    | Label l when pred.(pc) = [pc-1] && succ.(pc-1) = [pc] ->
+      Remove 1
+    | _ ->
+      Unchanged
   in
+  let inp = Analysis.as_analysis_input func version in
   match change_instrs transform inp with
+  | Some instrs ->
+    let version = { version with instrs; has_checkpoints = false } in
+    Some (Instr.replace_active_version func version)
+  | None ->
+    let version = { version with has_checkpoints = false } in
+    Some (Instr.replace_active_version func version)
+
+(* This takes the active version and duplicates it. Osr targets are
+ * updated to point to the currently active version *)
+let next_version (func:afunction) =
+  let cur_version = Instr.active_version func in
+  assert (cur_version.has_checkpoints);
+  let transform pc =
+    match[@warning "-4"] cur_version.instrs.(pc) with
+    | Osr {cond; target={func; version; label}; map} ->
+      let target = {func; version = cur_version.label; label} in
+      Replace (Osr {cond; target; map})
+    | _ -> Unchanged
+  in
+  let inp = Analysis.as_analysis_input func cur_version in
+  { label = Edit.fresh_version_label func cur_version.label;
+    instrs = (match change_instrs transform inp with | None -> cur_version.instrs | Some i -> i);
+    annotations = None;
+    has_checkpoints = true }
+
+let insert_assumption (func : afunction) osr_cond pc : version option =
+  let version = next_version func in
+  if (not version.has_checkpoints) then None else
+  let instrs = version.instrs in
+  let preds = Analysis.predecessors version.instrs in
+  (* Finds the highest up osr checkpoint in that basic block where the
+   * assumption can be placed. *)
+  let rec find_candidate_osr cond_vars pc acc =
+    if pc = 0 then acc else
+    match[@warning "-4"] instrs.(pc) with
+    | Osr _ -> find_candidate_osr cond_vars (pc-1) (Some pc)
+    | Label _ ->
+      begin match[@warning "-4"] preds.(pc) with
+      | [pred] -> find_candidate_osr cond_vars pred acc
+      | _ -> acc
+      end
+    | _ ->
+      assert (preds.(pc) = [pc-1]);
+      if Instr.independent instrs.(pc) osr_cond
+      then find_candidate_osr cond_vars (pc-1) acc
+      else acc
+  in
+  let osr_vars = expr_vars osr_cond in
+  begin match find_candidate_osr osr_vars (pc-1) None with
   | None -> None
-  | Some instrs -> Some (result_version func version.label instrs)
+  | Some pc ->
+    begin match[@warning "-4"] instrs.(pc) with
+    | Osr {cond; target; map} ->
+      instrs.(pc) <- Osr {cond=osr_cond::cond; target; map};
+      Some { version with instrs }
+    | _ -> assert (false)
+    end
+  end
+
+let hoist_assumption ({instrs} as inp) : instructions option =
```

Comment:
Maybe a comment that describes what this (non-trivial) function does from a high level could be helpful.